website/pages/blog/building-cloudquery.md

```diff
@@ -0,0 +1,104 @@
+
+We could take advantage of the schema definition and generate the source plugin from that, but we decided to do that one step further in the pipeline: from the Go Code.
+
+Go is incredibly suited for Code Analysis so we created a [code generation library](https://github.com/cloudquery/plugin-sdk/tree/main/codegen) that generates CloudQuery schema from Go structs. This helps achieve a number of things:
+
+1. Keep CloudQuery schema up-to-date with APIs automatically
+2. Enable developers to use the Go Client which is built for developers and include important capabilities when talking to the service API.
+3. Our current set of plugins are >80% auto-generated!
+
+## Scaling Destinations
+
+### Pluggable Destinations
+
+Another key requirement is supporting multiple destinations. The number of destinations is actually limited, but if destinations won't be decoupled from sources, source plugins will need to be updated every time there is a new destination, which will require a growing `n*m` (`n` number of source plugins and `m` number of destinations plugins) number of code-changes and make development unsustainable.
+
+Destination plugins are implemented in a similar way to source plugins as gRPC servers.
+
+### Typing System
+
+The job of a pluggable ELT platform is to do two things:
+
+1. Transform and normalize the data from an API
+2. For each transformed field, add type data.
+
+CloudQuery transforms every single field to its own rich type system, which contains more than [17 types](https://github.com/cloudquery/plugin-sdk/tree/main/schema) (including things like IP Addresses, MAC Addresses). This ensures all the validation is happening in the transformation phase. Destination plugins then only need transform this, depending on what types each destination supports. This is a big shift from how Singer or Airbyte works. These systems use what is available in JSON and [JSON Schema](https://json-schema.org/), as they couple the encoding together with the type system.
+
+Because of these two design decisions, CloudQuery already supports [5 destinations](https://www.cloudquery.io/docs/plugins/destinations) only two months after our initial [V1 release](https://www.cloudquery.io/blog/cloudquery-v1-release)
+
+## Performance
+
+Cloud Infrastructure hyperscale created new challenges for the ELT world. For example, some companies have more than 10,000 GCP and AWS accounts, with more than 100 million resources in total. How do you keep your inventory of these up-to-date, and fetch them on a daily (or twice daily) basis? To address this we made two architecture decisions:
+
+### Concurrency and Scheduling Model
+
+As part of our SDK, which provides the concurrency model for all the source plugins, we took advantage of the excellent concurrency support in Go. Go's goroutines enable us to scale to tens of thousands of API calls with a low number of OS threads and low memory overhead.
```

Comment:
Hmm, Go itself uses the term "goroutines" (lowercase): https://gobyexample.com/goroutines